syslog-ng-4.11.0 fails to build with openssl-4.0.0

# syslog-ng
## Version of syslog-ng
<!-- $ syslog-ng --version -->
syslog-ng 4 (4.11.0)
Config version: 4.2
Installer-Version: 4.11.0
Revision: 
Compile-Date: Mar 21 2026 11:33:50
Module-Directory: /usr/lib64/syslog-ng
Module-Path: /usr/lib64/syslog-ng
Include-Path: /usr/share/syslog-ng/include
Available-Modules: secure-logging,pseudofile,syslogformat,graphite,linux-kmsg-format,affile,metrics-probe,tfgetent,afstomp,map-value-pairs,tags-parser,cef,xml,hook-commands,cryptofuncs,afuser,regexp-parser,afprog,add-contextual-data,confgen,json-plugin,examples,system-source,rate-limit-filter,kvformat,disk-buffer,correlation,csvparser,stardate,afsocket,azure-auth-header,basicfuncs,timestamp,appmodel
Enable-Debug: off
Enable-GProf: off
Enable-Memtrace: off
Enable-Stackdump: off
Enable-IPv6: on
Enable-Spoof-Source: off
Enable-TCP-Wrapper: off
Enable-Linux-Caps: off
Enable-Systemd: off


## Platform
<!-- Name and version of OS -->
NAME='Gentoo'
ID='gentoo'
PRETTY_NAME='Gentoo Linux'
VERSION='2.18'

## Debug bundle
<!--
Create a debug bundle on your system with the syslog-ng-debun script which is included in the syslog-ng package.

Overview of the CLI options of syslog-ng-debun:
-r: run actual information gathering
-d: run syslog-ng in debug mode
-p: perform packet capture
-s: do strace
-t: timeout period for running debug/pcap/strace
-w: wait period before starting debug mode
-l: light information gathering (respects privacy)
-R: alternate installation directory for syslog-ng

$ syslog-ng-debun -r
-->
Syslog-NG DEBUg buNdle generator
Start environment detection
Linux-type FHS detected
Unknown or unsupported Linux distribution!

Operating System Name: Linux
Unknown Distro, perhaps unsupported
Unknown Distro, perhaps unsupported

Start general info collection
System's full uname: Linux gentoo-portage 6.17.9-1-pve #1 SMP PREEMPT_DYNAMIC PMX 6.17.9-1 (2026-01-12T16:25Z) x86_64 Intel(R) Core(TM) i7 CPU 860 @ 2.80GHz GenuineIntel GNU/Linux
Getting network-interface information: Success
Getting network routes: Success
Getting DNS resolution-related information: Done
List all processes
Mount and disk free info collection
RTNETLINK answers: Invalid argument
RTNETLINK answers: Invalid argument

Start Syslog-specific info collection
Copy configuration files from /etc/syslog-ng
/usr/bin/syslog-ng-debun: line 690: cpio: command not found
Copy SCL configuration files
/usr/bin/syslog-ng-debun: line 704: cpio: command not found
Gathering PKI information... done.
SVpid: 481 SNGpid: 482 Chpids: 
Syslog-ng's exact version: syslog-ng 4 (4.11.0)
/usr/sbin/syslog-ng-ctl stats
/usr/sbin/syslog-ng-ctl query get *
/usr/sbin/syslog-ng-ctl show-license-info
/usr/sbin/syslog-ng-ctl credentials status
/usr/bin/syslog-ng-debun: line 816: cpio: command not found
grep: warning: stray \ before /
Package file search is not supported (yet) on this platform
Package file search is not supported (yet) on this platform
Package file search is not supported (yet) on this platform
Package file search is not supported (yet) on this platform
Package file search is not supported (yet) on this platform
Package file search is not supported (yet) on this platform
Package file search is not supported (yet) on this platform
Package file search is not supported (yet) on this platform
Package file search is not supported (yet) on this platform
Package file search is not supported (yet) on this platform
Package file search is not supported (yet) on this platform
Package file search is not supported (yet) on this platform
Detecting init system: falling back to SystemV init style...
/usr/bin/syslog-ng-debun: line 278: kill: `none': not a pid or valid job spec
/usr/bin/syslog-ng-debun: line 281: kill: `none': not a pid or valid job spec
Generating second batch of statistics
/usr/sbin/syslog-ng-ctl stats
/usr/sbin/syslog-ng-ctl query get *

# Issue
## Failure

Can't build with openssl-4.0.0 package due to the following compilation error:
.....
lib/transport/tls-verifier.c: In function 'tls_verify_certificate_name':                                                                                                                      
lib/transport/tls-verifier.c:191:11: warning: assignment discards 'const' qualifier from pointer target type [-Wdiscarded-qualifiers]                                                         
  191 |       ext = X509_get_ext(cert, ext_ndx);                                                                                                                                              
      |           ^                                                                                                                                                                           
lib/transport/tls-verifier.c:224:44: error: invalid use of incomplete typedef 'ASN1_OCTET_STRING' {aka 'struct asn1_string_st'}                                                               
  224 |                   if (gen_name->d.iPAddress->length == 16)                                                                                                                            
      |                                            ^~                                                                                                                                         
lib/transport/tls-verifier.c:227:67: error: invalid use of incomplete typedef 'ASN1_OCTET_STRING' {aka 'struct asn1_string_st'}                                                               
  227 |                   if (inet_ntop(addr_family, gen_name->d.iPAddress->data, dotted_ip, sizeof(dotted_ip)))                                                                              
      |                                                                   ^~                                      

......
<!--

Backtrace, error messages or detailed description of failure comes here.

$ gdb syslog-ng
 run
-->

## Steps to reproduce
<!-- A description of how to trigger this bug. -->
Build always fail wit openssl-4.0.0

## Configuration
<!-- $ cat /path/to/syslog-ng.conf -->
@version: 4.11
##
## Syslog-ng default configuration file for Gentoo Linux

options {
	threaded(yes);
	chain_hostnames(no);

	## The default action of syslog-ng is to log a STATS line
	## to the file every 10 minutes.  That's pretty ugly after a while.
	## Change it to every 12 hours so you get a nice daily update of
	## how many messages syslog-ng missed (0).
	stats(freq(43200));
};

source src { system(); internal(); };

destination messages { file("/var/log/messages"); };

## By default messages are logged to tty12...
destination console_all { file("/dev/tty12"); };
## ...if you intend to use /dev/console for programs like xconsole
## you can comment out the destination line above that references /dev/tty12
## and uncomment the line below.
##destination console_all { file("/dev/console"); };

log { source(src); destination(messages); };
log { source(src); destination(console_all); };

## Input and output logs (if possible)

[syslog.debun.gentoo-portage.2026-04-22_12-23.205295.tgz](https://github.com/user-attachments/files/26962942/syslog.debun.gentoo-portage.2026-04-22_12-23.205295.tgz)

[build.log](https://github.com/user-attachments/files/26962967/build.log)

--- a/lib/transport/tls-verifier.c
+++ b/lib/transport/tls-verifier.c
@@ -220,10 +220,10 @@ tls_verify_certificate_name(X509 *cert, const gchar *host_name)
                 {
                   gchar dotted_ip[64] = {0};
                   int addr_family = AF_INET;
-                  if (gen_name->d.iPAddress->length == 16)
+                  if (ASN1_STRING_length(gen_name->d.iPAddress) == 16)
                     addr_family = AF_INET6;
 
-                  if (inet_ntop(addr_family, gen_name->d.iPAddress->data, dotted_ip, sizeof(dotted_ip)))
+                  if (inet_ntop(addr_family, ASN1_STRING_get0_data(gen_name->d.iPAddress), dotted_ip, sizeof(dotted_ip)))
                     {
                       g_strlcpy(pattern_buf, dotted_ip, sizeof(pattern_buf));
                       found = TRUE;

